mqtt/mqtt-bridge/src/persist/memory/loader.rs

```diff
@@ -0,0 +1,328 @@
+        loader.next().await.unwrap();
+        loader.next().await.unwrap();
+
+        // remove inserted elements
+        let mut state_lock = state.lock();
+        state_lock.remove_in_flight(&key1).unwrap();
+        state_lock.remove_in_flight(&key2).unwrap();
+        drop(state_lock);
+
+        // insert new elements
+        let key3 = Key { offset: 2 };
+        let pub3 = Publication {
+            topic_name: "test".to_string(),
+            qos: QoS::ExactlyOnce,
+            retain: true,
+            payload: Bytes::new(),
+        };
+        let mut state_lock = state.lock();
+        state_lock.insert(key3.clone(), pub3.clone());
+        drop(state_lock);
+
+        // verify new elements are there
+        let extracted = loader.next().await.unwrap();
+        assert_eq!(extracted.0, key3);
+        assert_eq!(extracted.1, pub3);
+    }
+
+    #[tokio::test]
+    async fn ordering_maintained_across_inserts() {
+        // setup state
+        let state = WakingMap::new();
+        let state = Arc::new(Mutex::new(state));
+
+        // add many elements
+        let mut state_lock = state.lock();
+        let num_elements = 50 as usize;
+        for i in 0..num_elements {
+            let key = Key { offset: i as u32 };
+            let publication = Publication {
+                topic_name: "test".to_string(),
+                qos: QoS::ExactlyOnce,
+                retain: true,
+                payload: Bytes::new(),
+            };
+
+            state_lock.insert(key, publication)
+        }
+
+        // verify insertion order
+        let elements: Vec<_> = get_elements(&mut state_lock, num_elements).collect();
+        for count in 0..num_elements {
+            assert_eq!(elements.get(count).unwrap().0.offset, count as u32)
+        }
+    }
+
+    #[tokio::test]
+    async fn poll_stream_does_not_block_when_map_empty() {
+        // setup state
+        let state = WakingMap::new();
+        let state = Arc::new(Mutex::new(state));
+
+        // setup data
+        let key1 = Key { offset: 0 };
+        let pub1 = Publication {
+            topic_name: "test".to_string(),
+            qos: QoS::ExactlyOnce,
+            retain: true,
+            payload: Bytes::new(),
+        };
+
+        // get loader
+        let batch_size = 5;
+        let mut loader = InMemoryMessageLoader::new(Arc::clone(&state), batch_size).await;
+
+        // async function that waits for a message to enter the state
+        let key_copy = key1.clone();
+        let pub_copy = pub1.clone();
+        let poll_stream = async move {
+            let maybe_extracted = loader.next().await;
+            if let Some(extracted) = maybe_extracted {
+                assert_eq!((key_copy, pub_copy), extracted);
+            }
+        };
+
+        // start the function and make sure it starts polling the stream before next step
+        let poll_stream_handle = tokio::spawn(poll_stream);
+        time::delay_for(Duration::from_secs(2)).await;
```

Comment:
I don't think we can use notify here because we would need to notify in the stream implementation